### PR TITLE
Allow manual setting of item repair cost

### PR DESCRIPTION
**Pull request by BlackxSnow**
#### Summary
Features "Allow manual setting of item repair cost"

#### Purpose of change
This change is useful in particular for modding, providing a way to bypass the automatic volume based repair calculation. This calculation increases linearly with volume, meaning that some items (such as bags) are prohibitively expensive to repair.

#### Describe the solution
Adding the option for a JSON field "repair_cost" stating the absolute material cost of repairing a point of HP.

#### Describe alternatives you've considered
Other viable solutions involve modifying the repair calculation. I do not think these solutions are mutually exclusive to this PR.
The first is to base the repair cost off of the initial crafting recipe cost, and the other is to modify the calculation to be based on the surface area, treating the object as either a sphere or a rectangular prism using the longest side.

#### Testing
Loaded game with a single item using the field, and no others changed. Field worked correctly on the item it was set on, and other items were still able to be repaired as normal using the volume calculation.

### CODE REVIEW

**Comment by kevingranade** [comment]
You need to run astyle over your changes, they don't match project style.
The various warnings this triggers need to be addressed.

But most importantly, what is the use case for this change?  In what scenarios is the calculated repair cost wrong *enough* that it needs an override?

**Comment by BlackxSnow** [comment]
> You need to run astyle over your changes, they don't match project style.
> The various warnings this triggers need to be addressed.
> 
> But most importantly, what is the use case for this change? In what scenarios is the calculated repair cost wrong _enough_ that it needs an override?

The high-volume rucksack takes 30 leather patches to make, and 9 to repair. That means repairing it from XX to full takes 36 patches. This is both stupidly impractical from a gameplay perspective and unrealistic even if you consider XX to be a fallen apart item (which it isn't, because weapons drop into components on destruction, not XX).

This was the exact scenario that made me create the PR in the first place. It also has no impact on the game unless specified in JSON so there's no chance of unintended imbalance.

**Comment by anothersimulacrum** [comment]
I feel as though we should improve the calculation, not add ad-hoc overrides throughout.

**Comment by Zireael07** [comment]
@anothersimulacrum: That's what was already tried (I think it used weight originally not volume) but there are always outliers where it doesn't make sense. I don't think it's possible to get a perfect calculation anymore...

**Comment by BlackxSnow** [comment]
@anothersimulacrum , @Zireael07 is correct. You will never find a mathematical equation that will perfectly represent a sane cost for repair on every item.

Besides, this change is: light on code, light on computation, and light on memory. It also provides another tool for modders to fine tune their additions rather than relying on clunky, hardcoded algorithms to determine something that really is a creative decision by the item author (for base DDA obviously adhering to reality).

**Comment by anothersimulacrum** [comment]
I do not think the impossibility to get a perfect formula should mean we should skip attempting to improve it, which is what this feels like.

**Comment by BlackxSnow** [comment]
@anothersimulacrum if you want to improve it, be my guest. I explicitly said that this PR is not incompatible with changing the formula, but I'm not interested in taking responsibility for balancing the hundreds of items and their repair costs in relation to a repair formula. :)

As I said, this PR has no real downsides, and is a relatively tiny change to the code base.

**Comment by anothersimulacrum** [comment]
I do not think this is problematic to have (for e.g. mods), but if you're wanting to apply it to any vanilla items, it is the wrong approach.

File: src/item_factory.cpp [hunk 1]
```diff
@@ -2917,6 +2917,8 @@ void Item_factory::load_basic_info( const JsonObject &jo, itype &def, const std:
         jo.read( "repairs_like", def.repairs_like );
     }
 
+    jo.read("repair_cost", def.repair_cost, false);
```
[the remark this answers is not in the document]
Comment (BlackxSnow) [reply src/item_factory.cpp:2920]: Both the getter function in item.h, and the field in itype.h are documented. No other "jo.read()" call is commented in this file - why should this be?

**Comment by BlackxSnow** [comment]
> I do not think this is problematic to have (for e.g. mods), but if you're wanting to apply it to any vanilla items, it is the wrong approach.

Outside of my own personal JSON modifications I have no particular interest in adding this field to vanilla items. I'm running this code in my own game so I can adjust repair costs to what I feel comfortable with, the PR is only because I figure I can't be the only one having issues with repair costs.

File: src/item_factory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (anothersimulacrum) [reply src/item_factory.cpp:2920]: Sorry, documentation for people not reading the C++, in doc/.

**Comment by kevingranade** [comment]
This is the wrong fix, it is definitely doable to adjust the repair cost to be bounded by the construction cost of the target item, and encouraging fixups just creates busywork.

**Pull request closed without merging**